In [0]:
import pandas as pd
data = pd.read_csv("10percent_sparse_smalldata.csv")

In [2]:
data

userId  movieId  rating     timestamp
0        1      100     3.5  1.260759e+09
1        2      100     5.0  8.353555e+08
2        3      100     2.0  8.353557e+08
3        4      100     3.0  8.353556e+08
4        5      100     3.0  8.353562e+08
..     ...      ...     ...           ...
95       1      119     4.0  1.260759e+09
96       2      119     5.0  8.353556e+08
97       3      119     3.5  8.353556e+08
98       4      119     3.0  8.353557e+08
99       5      119     3.0  1.298862e+09

[100 rows x 4 columns]

In [6]:
pivoted_data = data.pivot_table(index='userId', columns='movieId', values='rating')
pivoted_data

movieId  100  101  102  103  104  105  106  ...  113  114  115  116  117  118  119
userId                                      ...                                   
1        3.5  5.0  3.0  4.0  NaN  3.0  5.0  ...  4.0  2.0  4.0  5.0  4.0  4.0  4.0
2        5.0  NaN  5.0  3.0  4.0  2.0  3.5  ...  4.0  4.0  1.0  4.0  4.0  4.0  5.0
3        2.0  NaN  4.0  3.0  4.5  4.0  5.0  ...  5.0  3.0  3.0  3.5  5.0  4.0  3.5
4        3.0  3.5  NaN  5.0  4.0  4.0  4.0  ...  4.0  4.0  3.0  5.0  4.5  3.0  3.0
5        3.0  5.0  2.0  4.0  3.0  3.0  4.5  ...  NaN  5.0  4.0  NaN  5.0  4.0  3.0

[5 rows x 20 columns]

In [10]:
# use isna() function to detect which entries in a DataFrame are NA
na_rating = pivoted_data.isna()
na_rating

movieId    100    101    102    103    104  ...    115    116    117    118    119
userId                                      ...                                   
1        False  False  False  False   True  ...  False  False  False  False  False
2        False   True  False  False  False  ...  False  False  False  False  False
3        False   True  False  False  False  ...  False  False  False  False  False
4        False  False   True  False  False  ...  False  False  False  False  False
5        False  False  False  False  False  ...  False   True  False  False  False

[5 rows x 20 columns]

It return boolean for indicate which entries are NA

In [12]:
data_dropped_na = pivoted_data.dropna()
data_dropped_na

Empty DataFrame
Columns: [100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119]
Index: []

It will remove entire entries if remove which consists of NA, so it is not fit to our application. Next we need to check total number of NaN entry

In [20]:
number_NaN = pivoted_data.isna().sum()
number_NaN

movieId
100    0
101    2
102    1
103    0
104    1
105    0
106    0
107    1
108    1
109    0
110    1
111    0
112    1
113    1
114    0
115    0
116    1
117    0
118    0
119    0
dtype: int64

In [22]:
na_count = (len(pivoted_data)-pivoted_data.count()).sum()
na_count

10

In [36]:
# Check the dimension of data
pivoted_data.shape

(5, 20)

In [37]:
type(pivoted_data.shape)

tuple

In [39]:
number_value = pivoted_data.shape[0] * pivoted_data.shape[1]
number_value

100

In [40]:
percentage_of_sparsity = (na_count / number_value) * 100
percentage_of_sparsity

10.0

Therefore, the sparsity of dataset in percentage is 10% which is correct since the dataset is fabricated with 10% sparsity in 100 rating values. To deal with sparsity, several study need to conduct, we also can try to replace NaN entries by other values such as '0'.

In [43]:
filled_zero_data = pivoted_data.fillna(0)
filled_zero_data

movieId  100  101  102  103  104  105  106  ...  113  114  115  116  117  118  119
userId                                      ...                                   
1        3.5  5.0  3.0  4.0  0.0  3.0  5.0  ...  4.0  2.0  4.0  5.0  4.0  4.0  4.0
2        5.0  0.0  5.0  3.0  4.0  2.0  3.5  ...  4.0  4.0  1.0  4.0  4.0  4.0  5.0
3        2.0  0.0  4.0  3.0  4.5  4.0  5.0  ...  5.0  3.0  3.0  3.5  5.0  4.0  3.5
4        3.0  3.5  0.0  5.0  4.0  4.0  4.0  ...  4.0  4.0  3.0  5.0  4.5  3.0  3.0
5        3.0  5.0  2.0  4.0  3.0  3.0  4.5  ...  0.0  5.0  4.0  0.0  5.0  4.0  3.0

[5 rows x 20 columns]

Or maybe can replace with mean value? However, it depends on the application.

In [50]:
# Even if we do not specify axis = 0, 
# the method will return the mean over 
# the index axis by default 
mean_data = pivoted_data.mean(axis = 0, skipna = True) 
mean_data

movieId
100    3.300
101    4.500
102    3.500
103    3.800
104    3.875
105    3.200
106    4.400
107    3.750
108    4.000
109    4.100
110    4.125
111    4.300
112    3.375
113    4.250
114    3.600
115    3.000
116    4.375
117    4.500
118    3.800
119    3.700
dtype: float64

In [51]:
filled_mean_data = pivoted_data.fillna(mean_data)
filled_mean_data

movieId  100  101  102  103    104  105  ...  114  115    116  117  118  119
userId                                   ...                                
1        3.5  5.0  3.0  4.0  3.875  3.0  ...  2.0  4.0  5.000  4.0  4.0  4.0
2        5.0  4.5  5.0  3.0  4.000  2.0  ...  4.0  1.0  4.000  4.0  4.0  5.0
3        2.0  4.5  4.0  3.0  4.500  4.0  ...  3.0  3.0  3.500  5.0  4.0  3.5
4        3.0  3.5  3.5  5.0  4.000  4.0  ...  4.0  3.0  5.000  4.5  3.0  3.0
5        3.0  5.0  2.0  4.0  3.000  3.0  ...  5.0  4.0  4.375  5.0  4.0  3.0

[5 rows x 20 columns]

We can noticed that the NaN values is replaced by respective mean value in by columns. For example, movieId 110 has NaN value with index userId 1, then it filled up with mean value of 4.125.

After filled up NaN value, we can try to normalize the data. First, we need to check the max and min value in data

In [61]:
maximum = filled_mean_data.max()
max_rating = maximum.max()
max_rating

5.0

In [62]:
minimum = filled_mean_data.min()
min_rating = minimum.min()
min_rating

1.0

In [63]:
norm_data = (filled_mean_data - min_rating)/(max_rating - min_rating)
norm_data

movieId    100    101    102   103      104  ...   115      116    117   118    119
userId                                       ...                                   
1        0.625  1.000  0.500  0.75  0.71875  ...  0.75  1.00000  0.750  0.75  0.750
2        1.000  0.875  1.000  0.50  0.75000  ...  0.00  0.75000  0.750  0.75  1.000
3        0.250  0.875  0.750  0.50  0.87500  ...  0.50  0.62500  1.000  0.75  0.625
4        0.500  0.625  0.625  1.00  0.75000  ...  0.50  1.00000  0.875  0.50  0.500
5        0.500  1.000  0.250  0.75  0.50000  ...  0.75  0.84375  1.000  0.75  0.500

[5 rows x 20 columns]